chore: button component

deboer-tim opened the pull request:
### What does this PR do?

We currently use PatternFly classes like pf-c-button, pf-m-primary, pf-c-button__icon, and pf-m-start to style our buttons. Although they're powerful and we have a similar design, they tie us to PatternFly and its choices, and weren't really meant for desktop apps or dark mode. There are a few minor differences from our original design:
- slightly more button rounding
- in dark mode the hover should be lighter, not darker
- disabled buttons should be lighter
- spinner should replace button icon when there's progress
and this is repeated styles that aren't in a component.

This change adds a new Button component with support for styling, icons, and an inProgress spinner. It adds tests, and switches the buttons for the main lists, most form pages, and the MessageBox.

This is just a start to use the buttons in some core areas; after this PR there would need to be one or two more to replace all other instances.

### Screenshot/screencast of this PR

<img width="513" alt="Screenshot 2023-07-25 at 3 25 00 PM" src="https://github.com/containers/podman-desktop/assets/19958075/8a92ac23-19ce-4e61-a84e-a5e07910b5f7">

### What issues does this PR fix or reference?

N/A

### How to test this PR?

Go to all of these places and confirm that the buttons behave correctly, including disabled and in-progress states where applicable.

benoitf reviewed (commented):
hi, I have some comments

some could be just follow-up PR

we specify the type of the button.
Could we expect that by default buttons are primary button (so no need to specify each time)
instead of assuming they're not primary

I would not say a ButtonType is with 'undefined' type. It always need a type. (close or cancel would be a type) 


then, do we need a slot for the icon or we could just provide an optional icon object like `icon={fasTrash}`

benoitf commented on packages/renderer/src/lib/ui/Spinner.svelte:
```diff
@@ -2,7 +2,7 @@
 </script>
 
-<i class="pf-c-button__progress">
+<i class="pf-c-button__progress {$$props.class || ''}" style="{$$props.style || ''}">
```
if it's just to support position-fixed maybe we just need an attribute

benoitf commented on packages/renderer/src/lib/ui/Button.svelte:
```diff
@@ -0,0 +1,52 @@
+  {#if $$slots.icon}
+    <div class="flex flex-row p-0 m-0 bg-transparent justify-center space-x-[4px]">
+      {#if inProgress}
+        <Spinner size="sm" style="position: relative" />
```
I suppose `relative` doesn't work on class with PF ?

deboer-tim replied:
Nope. I'd prefer to just 'make it work normally', but replacing the spinner is an issue for another day. I'll use your other suggestion of an attribute for now.

benoitf commented on packages/renderer/src/lib/ui/Button.svelte:
```diff
@@ -0,0 +1,52 @@
+
+let classes = '';
+$: {
+  if (disabled || inProgress) {
+    if (type === 'primary' || type === 'secondary') {
+      classes = 'bg-charcoal-50';
+    } else {
+      classes = 'text-charcoal-50 no-underline';
+    }
+  } else {
+    if (type === 'primary') {
+      classes = 'bg-purple-500 border-none hover:bg-purple-400';
+    } else if (type === 'secondary') {
+      classes = 'border-[1px] border-gray-200 hover:border-purple-500 hover:text-purple-500';
+    } else {
+      classes = 'border-none hover:underline';
```
it makes me think, do we need `<button>` or `<primary-button>` `<secondary-button>` etc

deboer-tim replied:
Since most are primary, I think it'll be fine if we make that the default.

benoitf commented on packages/renderer/src/lib/ui/Button.ts:
```diff
@@ -0,0 +1,25 @@
+ *  secondary - a secondary action
+ *  undefined - close, cancel, or other non-default button
+ */
+export type ButtonType = 'primary' | 'secondary' | undefined;
```
instead of undefined it should be probably 'Link'

benoitf replied:
or 'link' (I have no preference between all lowercase, etc)

deboer-tim commented:
> we specify the type of the button. Could we expect that by default buttons are primary button (so no need to specify each time) instead of assuming they're not primary
>
> I would not say a ButtonType is with 'undefined' type. It always need a type. (close or cancel would be a type)

We have the three button types in ButtonType (primary, secondary, 'other'/links). I went with the patternfly default of only naming the first two, but you're right, it is the most common and should be default option. I can make primary the default and give the third one a name.

> then, do we need a slot for the icon or we could just provide an optional icon object like `icon={fasTrash}`

Yeah, I could do the same as EmptyScreen here.

benoitf commented:
for icons I'm also fine to have follow-up PRs

deboer-tim commented:
> > then, do we need a slot for the icon or we could just provide an optional icon object like `icon={fasTrash}`
> 
> Yeah, I could do the same as EmptyScreen here.

I went with the `button` default behaviour where title=tooltip and the text content is get button text, but as I remove the slot icons its odd that the text is the only thing that's not an attribute. Instead of `<Button title="This button does stuff">Click me</Button>`what do you think of `<Button title="Click me" tooltip="This button does stuff"/>`?

benoitf commented:
To me, I think `<Button title="This button does stuff">Click me</Button>` is more natural as it's close to what I would do with existing html component.

benoitf commented:
Then we would probably need to add https://sveltejs.github.io/eslint-plugin-svelte/rules/no-restricted-html-elements/ and say <button> is disallowed

deboer-tim commented:
Responded to all feedback with a new commit:
- Changed default button type to primary and give third option a name ('link')
- Made spinner support a boolean property to keep it simpler
- Changed button icon to a component using the same logic as EmptyScreen

Updated all uses of Button and tests to support the changes.

Note the one ugliness on Button.svelte:18 - we need to know when an icon like PodIcon is used because it has some required properties, and setting them on other components would produce a log warning like `<KubePlayIcon> was created with unknown prop 'solid'`. This isn't great but I couldn't find a simpler way to do it and IMHO still much better overall.

benoitf commented on packages/renderer/src/lib/ui/Button.svelte:
```diff
@@ -0,0 +1,71 @@
+onMount(() => {
+  if (icon?.prefix === 'fas') {
+    iconType = 'fa';
+  } else if (icon?.name.includes('PodIcon')) {
+    iconType = 'pd';
```
I'm wondering if we should not change defaults of PodIcon to be solid/1em size by default
so you wouldn't need to handle specifically PodIcon in buttons

deboer-tim replied:
I was debating that, but if we do it we should really change default for the matching container/image/volume icon too (although I don't think we don't use them on buttons yet, they are identical pattern and 1em solid seems like a reasonable default..).

Trying to limit the blast radius of this PR. Ok if I took that as separate cleanup-PR?

benoitf replied:
yes sure it's another PR

benoitf commented on packages/renderer/src/lib/ui/ButtonSpec.svelte:
```diff
@@ -0,0 +1,6 @@
+<script lang="ts">
+import { faLightbulb } from '@fortawesome/free-solid-svg-icons';
+import Button from './Button.svelte';
+</script>
+
+<Button title="Tooltip" icon="{faLightbulb}">Click me</Button>
```
I'm wondering why we have this component ?

couldn't we use icon property in the renderer and only in Button.spec.ts ?

benoitf replied:
I suppose you only need this component for the slot (name to be displayed in the button) but for the icon we don't

deboer-tim replied:
Yes, that's originally why it's there. It should be possible to move the icon to regular test now, I just didn't bother since we'd still need this unless there's a way to test the default slot, and it felt slightly less mock-y to me to have a component use icon={}? Could move it though, would be nicer if we could merge entirely.

benoitf reviewed (commented):
I'll perform UI check tomorrow

benoitf reviewed (approved):
checked UI and works 👍 
comments can be addressed as follow-up PRs
